### PR TITLE
Feat/global step synchronization

### PR DESCRIPTION
**Pull request by distributedstatemachine**
# **Changes**

### Global Step Synchronization

- **Embedded `global_step` in Model Slices:**
  - Modified `upload_slice_for_window` in `comms.py` to include `global_step` when uploading model slices (`state` and `delta`).
  - This ensures that each node shares its current training step with others through the uploaded slices.

- **Global Step Update from Applied Slices:**
  - Updated `apply_slices_to_model` to extract `global_step` from received slices.
  - Nodes update their local `global_step` to the maximum value among all received slices.
  - Synchronizes training progress across all miners and validators, preventing discrepancies.

### Checkpoint Loading and Saving

- **Proper Restoration of `global_step`:**
  - Ensured that `global_step` is correctly loaded from checkpoints during initialization.
  - Removed any redundant reinitialization that could reset `global_step` after loading.
  - Guarantees that nodes resume training or validation from the correct step count.

- **Saving `global_step` in Checkpoints:**
  - Updated checkpoint saving functions to include `global_step` in the saved state.
  - Enhances state persistence, allowing seamless continuation after restarts.

### Implementation Details

- **Modified `upload_slice_for_window` Function in `src/templar/comms.py`:**

  ```python
  async def upload_slice_for_window(
      bucket: str,
      model: torch.nn.Module,
      window: int,
      seed: str,
      wallet: 'bt.wallet',
      compression: int,
      key: str = 'slice',
      global_step: int = 0
  ):
      filename = f'{key}-{window}-{wallet.hotkey.ss58_address}.pt'
      logger.debug(f"Uploading slice to S3: {filename}")

      # Get indices for slicing
      indices = await get_indices_for_window(model, seed, compression)

      # Create the slice data with global_step
      slice_data = {'global_step': global_step}
      for name, param in model.named_parameters():
          slice_data[name] = param.data.view(-1)[indices[name].to(model.device)].cpu()

      # Save and upload the slice_data
      # ... existing code to save and upload to S3 ...
  ```

- **Updated `apply_slices_to_model` Function in `src/templar/comms.py`:**

  ```python
  async def apply_slices_to_model(
      model: torch.nn.Module,
      window: int,
      seed: str,
      compression: int,
      key: str = 'slice'
  ) -> int:
      indices_dict = await get_indices_for_window(model, seed, compression)
      slice_files = await load_files_for_window(window=window, key=key)

      # Initialize max_global_step
      max_global_step = 0

      # Iterate over each slice file
      for file_i in slice_files:
          try:
              slice_i = await get_slices(file_i, model.device)
              slice_global_step = slice_i.get('global_step', 0)  # Default to 0 if not present
              max_global_step = max(max_global_step, slice_global_step)

              # Apply the slice to the model
              # ... existing code to apply parameter slices ...
          except Exception as e:
              logger.exception(f"Error applying slice from {file_i}: {e}")

      # Return the maximum global_step found
      return max_global_step
  ```

- **Adjusted Miner and Validator Initialization:**

  - **Loading Checkpoint in `neurons/miner.py` and `neurons/validator.py`:**

    ```python
    # Load checkpoint if it exists
    if os.path.exists(self.checkpoint_path):
        tplr.logger.info(f"Loading checkpoint from {self.checkpoint_path}")
        global_step, _ = asyncio.run(load_checkpoint(
            filename=self.checkpoint_path,
            model=self.model,
            optimizer=self.optimizer,  # For miners
            scheduler=None,            # Scheduler initialized after loading
            device=self.config.device
        ))
        self.global_step = global_step  # Properly restore global_step
        tplr.logger.info(f"Resumed from global step {self.global_step}")
    else:
        tplr.logger.info("No checkpoint found. Starting from scratch.")
        self.global_step = 0
    ```

  - **Initializing Learning Rate Scheduler with Correct `global_step`:**

    ```python
    # Set last_epoch to global_step - 1
    self.scheduler = tplr.get_wsd_scheduler(
        optimizer=self.optimizer,
        num_warmup_steps=self.hparams.num_warmup_steps,
        num_stable_steps=self.hparams.num_stable_steps,
        num_decay_steps=self.hparams.num_decay_steps,
        last_epoch=self.global_step - 1
    )
    ```

- **Updating `global_step` After Applying Slices:**

  ```python
  # After applying slices, update local global_step
  max_global_step = await tplr.apply_slices_to_model(
      model=self.model,
      window=window,
      seed=window,
      compression=self.hparams.compression,
      key='state'
  )
  if max_global_step > self.global_step:
      self.global_step = max_global_step
      self.scheduler.last_epoch = self.global_step - 1  # Update scheduler
      tplr.logger.info(f"Updated global step to {self.global_step}")
  ```

**By implementing these changes:**

- **Consistent Training Progress:**
  - Nodes synchronize their `global_step`, ensuring they are at the same point in training.
  - New joiners and restarts seamlessly catch up with the network's training progress.

- **Unified Learning Rate Scheduling:**
  - All nodes transition through learning rate phases simultaneously.
  - Prevents some nodes from being in the warm-up phase while others are in decay, which could cause optimization conflicts.

- **Improved Model Convergence:**
  - Synchronization reduces conflicting updates and enhances collaborative training.
  - Leads to more stable and efficient convergence of the model.

- **Robust Restart Handling:**
  - Nodes resume accurately from their last state after interruptions.
  - Avoids loss of progress and redundant computations.

### Documentation

- **Added `docs/global_sync.md`:**
  - Provides a comprehensive explanation of the global step synchronization mechanism.
  - Details the motivation, implementation steps, and benefits.
  - Serves as a reference for developers and contributors to understand the synchronization process.

### CODE REVIEW

File: docs/global_sync.md [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+    global_step: int = 0
+):
+    filename = f'{key}-{window}-{wallet.hotkey.ss58_address}.pt'
+    logger.debug(f"Uploading slice to S3: {filename}")
+
+    # Get indices for slicing
+    indices = await get_indices_for_window(model, seed, compression)
+
+    # Create the slice data with global_step
+    slice_data = {'global_step': global_step}
```
Comment (noaonoszko) [inline docs/global_sync.md:53]: Nitpick: Not sure it makes sense do mix `global_step` and parameter layers on the same level in a dict, I would prefer something like
```python
slice_data = {"global_step": global_step, "state_dict": state_dict}
```

**Review by noaonoszko — COMMENTED** [review]
Looks great! However I suggest we remove the global sync docs completely. Currently it is only the lr schedulers that are synced, so I think we're better off just adding a sentence about this somewhere else in the docs to avoid cognitive overload for our dear readers and avoid having outdated docs since `global_sync.md` contains  code.

In addition to this I made two small nit picks on the code.

File: docs/global_sync.md (hunk 1 above)
Comment (noaonoszko) [reply docs/global_sync.md:53]: We could even make it a pydantic object to validate that `global_step`, `state_dict` and any future keys are present. Not necessary in this PR though.

Comment (distributedstatemachine) [reply docs/global_sync.md:53]: This will be the next PR . I plan on adding more miner metadata to it .
